webrtc/README.md

```diff
@@ -0,0 +1,445 @@
+   have a rate limiting mechanism in place as a defense measure. See also
+   https://datatracker.ietf.org/doc/html/rfc5389#section-16.1.2.
+
+7. _A_ and _B_ execute the DTLS handshake as part of the standard WebRTC
+   connection establishment.
+
+   At this point _B_ does not know the TLS certificate fingerprint of _A_. Thus
+   _B_ can not verify _A_'s TLS certificate fingerprint during the DTLS
+   handshake. Instead _B_ needs to _disable certificate fingerprint
+   verification_ (see e.g. [Pion's `disableCertificateFingerprintVerification`
+   option](https://github.com/pion/webrtc/blob/360b0f1745c7244850ed638f423cda716a81cedf/settingengine.go#L62)).
+
+   On success of the DTLS handshake the connection provides confidentiality and
+   integrity but not authenticity. The latter is guaranteed through the
+   succeeding Noise handshake. See [Connection Security
+   section](#connection-security).
+
+8. Messages on the established `RTCDataChannel` are framed using the message
+   framing mechanism described in [Multiplexing](#multiplexing).
+
+9. The remote is authenticated via an additional Noise handshake. See
+   [Connection Security section](#connection-security).
+
+WebRTC can run both on UDP and TCP. libp2p WebRTC implementations MUST support
+UDP and MAY support TCP.
+
+## Multiplexing
+
+The WebRTC browser APIs do not support half-closing of streams nor resets of the
+sending part of streams.
+[`RTCDataChannel.close()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close)
+flushes the remaining messages and closes the local write and read side. After
+calling `RTCDataChannel.close()` one can no longer read from the channel. This
+lack of functionality is problematic, given that libp2p protocols running on top
+of transport protocols, like WebRTC, expect to be able to half-close or reset a
+stream. See [Connection Establishment in
+libp2p](https://github.com/libp2p/specs/blob/master/connections/README.md#definitions).
+
+To support half-closing and resets of streams, libp2p WebRTC uses message
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and sent on the `RTCDataChannel` prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
```

Comment:
This is something that would need to be confirmed, but I believe that `prost`, the library used by rust-libp2p, isn't capable of differentiating between missing fields and zeroed out fields. When `prost` decodes a message, any missing field is replaced with a 0 value. Thus I believe that it's not capable of differentiating between "no flags" and "fin".

Reply:
Since a `1`, `2`, or other doesn't cost any less or more bytes or CPU than a `0`, this could easily be bypassed by having for example `FIN = 3`.

Reply:
Sounds reasonable if that’s how the protobuf parser behaves. If I remember correctly, @mxinden once mentioned that it does the correct thing though. @mxinden, can you double-check?

Reply:
The `flag` field is represented as an `Option<_>` in the generated Rust code, see e.g.:

https://github.com/libp2p/rust-libp2p/blob/8ddd225e26f6fd08a1ab16035eabf5b700cdffec/transports/webrtc/src/tokio/substream.rs#L96-L99

Reply:
I raised this because I've recently detected a problem in Substrate where a field value was not present in the protobuf message that Substrate sends out when the value of that field would have been 0. Just like here, the field was an enum.
See https://github.com/paritytech/substrate/pull/12438
You're confident that it works, I trust your judgement.

Reply:
For a full discussion, and likely also a pointer to the issue you are seeing in Substrate see https://github.com/libp2p/specs/issues/465.

Reply:
Resolving here. Let us know if you have related questions @tomaka.